test: Redeploy DNS after endpointRoutes reconfiguration

joestringer opened the pull request:
In general up until now, Cilium has expected endpointRoutes mode to be
set to exactly one value upon deployment and for that value to stay the
same for the remainder of operation. Toggling it can lead to a mix of
endpoints in different datapath modes which is not well covered in CI.

In Github issue #16717 we observed that if the testsuite toggles this
setting then we can end up with kubedns pods remaining in endpoint
routes mode, even though the rest of the daemon (and other pods) are not
configured in this mode. This can lead to connectivity issues in DNS,
and a range of test failures in subsequent tests because DNS is broken.

Longer term to resolve this, we could improve on Cilium to ensure that
users can successfully toggle this setting on or off at runtime and
properly handle this case, or alternatively shift all logic over to
endpoint-routes mode by default and disable the other option.

Given that CI for the master branch is in a poor state due to this issue
today, and that part of the issue is CI reconfiguring the datapath state
of Cilium during the test setup in an unsupported manner, this commit
proposes to force DNS pod redeployment as part of setup any time a test
reconfigures the endpointRoutes mode. This should mitigate the testing
side issue while we mull over the right longer-term solution.

Fixes: #16717
Fixes: https://github.com/cilium/cilium/pull/16227
Related: #14955

pchaigno commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -318,10 +323,14 @@ var _ = Describe("K8sDatapathConfig", func() {
+
+		// CAUTION: endpointRoutes mode was toggled. The next test
+		// MUST redeploy kubedns via RedeployCiliumAfterDatapathChange
+		// or potentially face GH-16717.
```
Couldn't we have a variable in `kubectl` to remember the current state of endpoint routes and use that to detect when we should restart DNS pods?

joestringer replied:
I briefly thought about this but discounted it with some hand-waving about storing state across tests being bad. Coming fresh after the weekend, I think this would be a much cleaner approach, so I'll try this out.

pchaigno replied:
> storing state across tests being bad

Yeah, I was concerned about that as well, but relying on people to call the correct code in the correct order is probably worse.

joestringer committed:
test: Redeploy DNS after changing endpointRoutes

In general up until now, Cilium has expected endpointRoutes mode to be
set to exactly one value upon deployment and for that value to stay the
same for the remainder of operation. Toggling it can lead to a mix of
endpoints in different datapath modes which is not well covered in CI.

In Github issue #16717 we observed that if the testsuite toggles this
setting then we can end up with kubedns pods remaining in endpoint
routes mode, even though the rest of the daemon (and other pods) are not
configured in this mode. This can lead to connectivity issues in DNS,
and a range of test failures in subsequent tests because DNS is broken.

Longer term to resolve this, we could improve on Cilium to ensure that
users can successfully toggle this setting on or off at runtime and
properly handle this case, or alternatively shift all logic over to
endpoint-routes mode by default and disable the other option.

Given that CI for the master branch is in a poor state due to this issue
today, and that part of the issue is CI reconfiguring the datapath state
of Cilium during the test setup in an unsupported manner, this commit
proposes to force DNS pod redeployment as part of setup any time a test
reconfigures the endpointRoutes mode. This should mitigate the testing
side issue while we mull over the right longer-term solution.

Signed-off-by: Joe Stringer <joe@cilium.io>

pchaigno reviewed (approved):
New version LGTM as well.
